src/dataflow/analysis/tests/analysis-test.ts

```diff
@@ -28,6 +29,245 @@ function assertFailures(result: ValidationResult, expectedFailures: string[]) {
   assert.sameMembers([...result.failures], [...expectedFailuresWithoutPaths]);
 }
 
+/**
+ * Creates test nodes and edges in a chain, using the given node IDs. e.g.
+ * createChainOfEdges('A', 'B', 'C') will return an array of two edges: AB, BC.
+ */
+function createChainOfEdges(...nodeIds: string[]) {
+  assert(nodeIds.length >= 2);
+  const nodes = nodeIds.map(nodeId => new TestNode(nodeId));
```

Comment:
Might be handy to handle repeated nodeIds here, but I'm aware that it's likely out of scope for this PR.
i.e. a map from nodeId to nodes so that repeated Ids map to the same node.
This will make it easy to generate cycles later.

Reply:
I was wondering what to do about cycles... I like this idea! But I'll wait until I need cycles before I add it.